AppController/djinn.rb

```diff
@@ -2516,200 +2503,112 @@ def start_roles_on_nodes(ips_hash, secret)
-    ips_to_roles.each { |ip, roles|
-      Djinn.log_info("Will add roles #{roles.join(', ')} to new " +
-        "node at IP address #{ip}")
-      nodes_info << {
-        "public_ip" => ip,
-        "private_ip" => ip,
-        "jobs" => roles,
-        "disk" => nil
+        # Save the roles we want for this specific IP.
+        found = false
+        node_roles.each { |node|
+          if node['private_ip'] == ip
+            node.jobs << role
+            found = true
+            break
+          end
+        }
+        unless found
+          node_roles << {
+            "public_ip" => ip,
+            "private_ip" => ip,
+            "jobs" => role,
+            "disk" => nil
+          }
+        end
       }
     }
 
-    add_nodes(nodes_info)
-    update_hosts_info()
-
-    return nodes_info
-  end
-
-
-  # Starts the given roles by using open nodes, spawning new nodes, or some
-  # combination of the two.
-  #
-  # Args:
-  #   nodes_needed:  An Array, where each item is an Array of the roles to
-  #     start on each node.
-  #   instance_type: A String with the type of instance to start.
-  #   secret: A String with the deployment secret key.
-  # Returns:
-  #  A String with 'OK' or the Error string.
-  def start_new_roles_on_nodes(nodes_needed, instance_type, secret)
-    return BAD_SECRET_MSG unless valid_secret?(secret)
-
-    # TODO: we should validate the roles type, to ensure they are valid.
-    if nodes_needed.class != Array || (instance_type.class != String && is_cloud)
-      Djinn.log_error("Invalid parameter type (nodes_needed, or instance_type).")
-      return BAD_INPUT_MSG
-    end
-
-    Djinn.log_info("Received a request to acquire nodes with roles " +
-      "#{nodes_needed.join(', ')}, with instance type #{instance_type}.")
-
-    vms_to_use = []
-
-    # We look for 'open' nodes first.
+    # Let's count the open nodes we can use before having to create new
+    # VMs.
+    open_nodes = 0
     @state_change_lock.synchronize {
-      @nodes.each { |node|
-        if node.is_open?
-          Djinn.log_info("Will use node #{node} to run new roles")
-          node.jobs = nodes_needed[vms_to_use.length]
-          vms_to_use << node
-        end
-      }
+      @nodes.each { |node| open_nodes += 1 if node.is_open?  }
     }
-    Djinn.log_info("Using #{vms_to_use.length} open nodes to run new roles.")
 
-    vms_to_spawn = nodes_needed.length - vms_to_use.length
-
-    if vms_to_spawn > 0
+    # We spawn new nodes if we need to (and can do so) here.
+    new_nodes_info = []
+    if new_nodes_roles.length > open_nodes
+      num_of_vms = new_nodes_roles.length - open_nodes
       unless is_cloud?
-        Djinn.log_error("Still need #{vms_to_spawn} more nodes, but we " +
+        Djinn.log_warn("Still need #{num_of_vms} more nodes, but we " +
         "aren't in a cloud environment, so we can't acquire more nodes - " +
         "failing the caller's request.")
         return NOT_ENOUGH_OPEN_NODES
```

Comment:
I'm not sure the exception is carried well across SOAP, and also that the tools handles it. I would say that if we go for it, we may want to wait for the Admin API and have this case handled there.